modules/auxiliary/gather/enum_dns.rb

```diff
@@ -1,483 +1,420 @@
+        dns.nameservers -= dns.nameservers
+        dns.nameservers = ("#{datastore['NS']}")
       end
-      return true
-    else
-      return false
+      dns.use_tcp = datastore['TCP_DNS']
+      dns.udp_timeout = datastore['TIMEOUT']
+      dns.retry_number = datastore['RETRY']
+      dns.retry_interval = datastore['RETRY_INTERVAL']
+      dns.query(domain, type)
+    rescue ::Rex::ConnectionError
+    rescue ::Rex::ConnectionRefused
+    rescue ::Rex::ConnectionTimeout
+    rescue ::Rex::SocketError
+    rescue ::Rex::TimeoutError
+    rescue ::Timeout::Error
     end
   end
 
-  def genrcd(target)
-    print_status("Retrieving general DNS records")
-    query = @res.search(target)
-    if (query)
-      query.answer.each do |rr|
-        next unless rr.class == Net::DNS::RR::A
-        print_status("Domain: #{target} IP address: #{rr.address} Record: A ")
-        report_note(:host => @nsinuse.to_s,
-          :proto => 'udp',
-          :sname => 'dns',
-          :port => 53 ,
-          :type => 'dns.enum',
-          :update => :unique_data,
-          :data => "#{rr.address.to_s},#{target},A")
-      end
-    end
-    query = @res.query(target, "SOA")
-    if (query)
-      (query.answer.select { |i| i.class == Net::DNS::RR::SOA}).each do |rr|
-        query1 = @res.search(rr.mname)
-        if (query1)
-          query1.answer.each do |ip|
-            print_status("Start of Authority: #{rr.mname} IP address: #{ip.address} Record: SOA")
-            report_note(:host => @nsinuse.to_s,
-              :proto => 'udp',
-              :sname => 'dns',
-              :port => 53 ,
-              :type => 'dns.enum',
-              :update => :unique_data,
-              :data => "#{ip.address.to_s},#{rr.mname},SOA")
-          end
-        end
-      end
+  def dns_bruteforce(domain)
+    wordlist = datastore['WORDLIST'].to_s
+    return if wordlist.blank?
+
+    threadnm = datastore['THREADS'].to_i
+    return if threadnm == 0
+
+    queue = ::Queue.new
+    File.foreach(wordlist) do |line|
+      queue << "#{line.chomp}.#{domain}"
     end
-    query = @res.query(target, "NS")
-    if (query)
-      (query.answer.select { |i| i.class == Net::DNS::RR::NS}).each do |rr|
-        query1 = @res.search(rr.nsdname)
-        if (query1)
-          query1.answer.each do |ip|
-            next unless ip.class == Net::DNS::RR::A
-            print_status("Name Server: #{rr.nsdname} IP address: #{ip.address} Record: NS")
-            report_note(:host => @nsinuse.to_s,
-              :proto => 'udp',
-              :sname => 'dns',
-              :port => 53 ,
-              :type => 'dns.enum',
-              :update => :unique_data,
-              :data => "#{ip.address.to_s},#{rr.nsdname},NS")
+
+    records = []
+    until queue.empty?
+      tl = []
+      1.upto(threadnm) do
+        tl << framework.threads.spawn(
+          "Module(#{refname})-#{domain}", false, queue.shift) do |target|
+            Thread.current.kill unless target
```

Comment:
This just seems weird -- why spawn and immediately kill a thread if there isn't a target? Why not `queue.shift` earlier and only spawn if a target is available?